dds/DCPS/DiscoveryBase.h

```diff
@@ -1263,6 +1231,244 @@ namespace OpenDDS {
         }
       }
 
+      void
+      remove_expired_endpoints(const DCPS::MonotonicTimePoint& /*now*/)
+      {
+        // TLS_TODO: sanity check on locks
+        ACE_GUARD(ACE_Thread_Mutex, g1, matching_data_buffer_lock_);
+        MatchingDataIter it;
+        for (it = matching_data_buffer_.begin(); it != matching_data_buffer_.end(); it++) {
+          if (MonotonicTimePoint::now() - it->second.time_added_to_map >= max_type_lookup_service_reply_period_) {
+            matching_data_buffer_.erase(it->first);
+          }
+        }
+        // TLS_TODO: any other cleanup?
+      }
+
+      void
+      match_continue(OpenDDS::DCPS::SequenceNumber rpc_sequence_number)
+      {
+        // TLS_TODO: sanity check on locks
+        // TLS_TODO: match_continue(it->second.writer, it->second.reader)
+        // will unlock '_lock'. Should it be locked here?
```

Comment:
I see two reverse locks, is that what you're talking about?

Reply:
Probably both... In one thread they get locked in match(), and unlocked in the 'lower level' match_continue(), which is also called from this function. The callers of this function (in Sedp) do not lock (at least I don't see that). So in this thread lower level match_continue() would unlock objects that were never locked. That's why I am locking them in this function. So I wanted to verify if this makes sense.

Reply:
Yea, I don't know about this one. I can say it does seem weird to grab a lock just for a function we're calling to unlock it. It doesn't help that `lock_` implies it's just one object-wide lock.

Reply:
Locking seems to be correct on review.  May remove secondary lock in the future.